python/ray/experimental/data/read_api.py

```diff
@@ -198,7 +201,73 @@ def read_json(paths: Union[str, List[str]],
     Returns:
         Dataset holding Arrow records read from the specified paths.
     """
-    raise NotImplementedError  # P0
+    import pyarrow as pa
+    from pyarrow.fs import (FileType, FileSelector,
+                            _resolve_filesystem_and_path)
+    from pyarrow import json
+    import numpy as np
+
+    # TODO(Clark): Refactor to use shared filesystem + path resolution utility
```

Comment:
Sounds good, we could move it into data source common code as well (make a datasource/ directory).